src/Data.Json.JTable.Internal.purs

```diff
@@ -0,0 +1,214 @@
+import qualified Data.Array.Unsafe as AU
+import qualified Data.StrMap as M
+import Data.Foldable (foldl, any, all, mconcat, elem)
+import Data.Traversable (for)
+import Control.Alt
+import Control.Apply
+import Control.Bind
+import Control.Monad
+import Control.MonadPlus
+
+import Data.Argonaut.Core
+import Data.Argonaut.JCursor
+import Data.Argonaut.Parser (jsonParser)
+import Text.Smolder.HTML (thead, tbody, tr, th, td)
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
+
+-- header data
+data Tree = T JPath Number Number [Tree]
+tPath (T p _ _ _) = p
+tWidth (T _ w _ _) = w
+tHeight (T _ _ h _) = h
+tKids (T _ _ _ k) = k
+
+instance showTree :: Show Tree where
+  show (T p w h k) = joinWith " " ["(T", show p, show w, show h, show k, ")"]
+
+-- cell data
+data Cell = C  JCursor Number Number JsonPrim
+cCursor (C c _ _ _) = c
+cWidth (C _ w _ _) = w
+cHeight (C _ _ h _) = h
+cJsonPrim (C _ _ _ j) = j
+
+instance showCell :: Show Cell where
+  show (C c w h j) = joinWith " " ["(C", show w, show h, show c, show j, ")"]
+
+
+foreign import jnull "var jnull = null;" :: Json
+
+_cN = const primNull
+toPrim = (foldJson _cN primBool primNum primStr _cN _cN) :: Json -> JsonPrim
+zipWithIndex xs = zip xs (0 .. length xs - 1) 
+
+so :: forall a. Boolean -> a -> Maybe a
+so b x = if b then Just x else Nothing
+
+
+-- maybe return the width of a tuple composed of primitive values
+widthOfPrimTuple :: JPath -> [Json] -> Maybe Number
+widthOfPrimTuple path ja = 
+  path !! 0 *> ja !! 1 *> let
+    types = ja <#> foldJson (\_->0) (\_->1) (\_->2) (\_->3) (\_->4) (\_->4)
+    not_same = length (nub types) /= 1
+    all_prim = all ((/=) 4) types
+    in (all_prim && (not_same || length ja == 2)) `so` length ja
+
+
+-- add child to tree, unify if exists
+tMergeArray :: Tree -> Tree -> Tree
+tMergeArray (T p w h k) nt@(T np nw nh nk) =
+  let i = findIndex (\n -> last np == last (n # tPath)) k in case k !! i of
+    Just t2@(T p2 w2 h2 k2) -> case foldl tMergeArray t2 nk of 
+      (T _ w2' h2' k2') -> let k' = updateAt i (T p2 w2' h2' k2') k
+                               w' = w - w2 + w2'
+                               h' = max h (h2' + 1)
+                           in T p w' h' k'
+    Nothing -> let w' = if null k then nw else w+nw
+                   h' = max h $ nh + 1
+                   k' = snoc k nt
+               in T p w' h' k'
+
+-- produce a tree of header data from json
+tFromJson :: JPath -> Json -> Tree
+tFromJson path json = let
+  obj = toObject json <#> \jo ->
+    if M.isEmpty jo then T path 1 0 []    
+    else let k = M.toList jo <#> uncurry \l j -> tFromJson (snoc path l) j
+             w = k <#> tWidth # foldl (+) 0
+             h = 1 + foldl max 0 (k <#> tHeight)
+         in T path w h k
+  m = (>>=) (toArray json)
+  tuple = m \a -> (widthOfPrimTuple path a) <#> \n -> T path n 0 []
```

Comment:
This code will have to change to detect and manage "vertically type compatible" tuples.